lib/psabpf_action_selector.c

```diff
@@ -728,3 +780,243 @@ uint32_t psabpf_action_selector_get_action_id_by_name(psabpf_action_selector_con
+        fprintf(stderr, "invalid map of groups\n");
+        return EINVAL;
+    }
+
+    /* Just validate that group exists */
+    uint32_t inner_map_id = 0;
+    int err = bpf_map_lookup_elem(ctx->map_of_groups.fd, &group->group_ref, &inner_map_id);
+    if (err != 0) {
+        err = errno;
+        fprintf(stderr, "failed to get group: %s\n", strerror(err));
+        return err;
+    }
+
+    return NO_ERROR;
+}
+
+psabpf_action_selector_group_context_t *psabpf_action_selector_get_next_group(psabpf_action_selector_context_t *ctx)
+{
+    if (ctx == NULL)
+        return NULL;
+    if (ctx->map_of_groups.key_size != 4) {
+        fprintf(stderr, "invalid map of groups\n");
+        return NULL;
+    }
+
+    close_object_fd(&ctx->group.fd);
+    psabpf_action_selector_group_free(&ctx->current_group);
+    psabpf_action_selector_group_init(&ctx->current_group);
+
+    if (ctx->current_group_id == PSABPF_ACTION_SELECTOR_INVALID_REFERENCE) {
+        /* group map is not open, so we start from this point */
+        if (bpf_map_get_next_key(ctx->map_of_groups.fd, NULL, &ctx->current_group.group_ref) != 0)
+            goto err_or_no_more_groups;
+    } else {
+        /* find next group reference */
+        if (bpf_map_get_next_key(ctx->map_of_groups.fd, &ctx->current_group_id, &ctx->current_group.group_ref) != 0)
+            goto err_or_no_more_groups;
+    }
+
+    ctx->current_group_id = ctx->current_group.group_ref;
+    return &ctx->current_group;
+
+err_or_no_more_groups:
+    ctx->current_group_id = PSABPF_ACTION_SELECTOR_INVALID_REFERENCE;
+    return NULL;
+}
+
+static int get_member_action(psabpf_action_selector_context_t *ctx,
+                             psabpf_action_selector_member_context_t *member)
+{
+    psabpf_table_entry_ctx_t tec = {
+            .table = ctx->map_of_members,
+            .btf_metadata = ctx->btf,
+            .cache = ctx->cache,
+    };
+    psabpf_match_key_t mk[] = {
+            {
+                    .type = PSABPF_EXACT,
+                    .key_size = sizeof(member->member_ref),
+                    .data = &member->member_ref,
+            },
+    };
+    psabpf_match_key_t * mk_ptr = &(mk[0]);
+    psabpf_table_entry_t te = {
+            .match_keys = &mk_ptr,
+            .n_keys = 1,
+    };
+
+    int ret = psabpf_table_entry_get(&tec, &te);
+    if (te.action != NULL) {
+        memcpy(&member->action, te.action, sizeof(psabpf_action_t));
+        free(te.action);
+    }
+
+    return ret;
+}
+
+psabpf_action_selector_member_context_t *psabpf_action_selector_get_next_group_member(psabpf_action_selector_context_t *ctx,
+                                                                                      psabpf_action_selector_group_context_t *group)
+{
+    if (ctx == NULL)
+        return NULL;
+    if (ctx->group.key_size != 4 || ctx->group.value_size != 4) {
+        fprintf(stderr, "invalid group map\n");
```

Comment:
Why this is invalid?

Reply:
For group map:
- key is a group reference, so it is always 32 bits wide;
- value is a reference to a member (or first entry is a number of references), so it is also 32 bits wide.